moshi/src/test/java/com/squareup/moshi/TypesTest.java

```diff
@@ -321,6 +321,40 @@ public static Type getFirstTypeArgument(Type type) throws Exception {
     }
   }
 
+  //
+  // Regression tests for https://github.com/square/moshi/issues/338
+  //
+  // Adapted from https://github.com/google/gson/pull/1128
+  //
+
+  private static final class RecursiveTypeVars<T> {
+    RecursiveTypeVars<? super T> superType;
+  }
+
+  @Test public void recursiveTypeVariablesResolve() {
+    JsonAdapter<RecursiveTypeVars<String>> adapter = new Moshi.Builder().build().adapter(Types
+        .newParameterizedTypeWithOwner(TypesTest.class, RecursiveTypeVars.class, String.class));
+    assertThat(adapter).isNotNull();
+  }
+
+  @Test public void testRecursiveTypeVariablesResolve1() {
```

Comment:
The other Moshi tests do not use the "testBlah" naming structure; they just do "blah."

Reply:
Part of dafa839